<!DOCTYPE html>
<!-- START: inst/pkgdown/templates/layout.html --><!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><title>Introduction to data analysis with R and Bioconductor: Manipulating and analysing data with dplyr</title><meta name="viewport" content="width=device-width, initial-scale=1"><link rel="stylesheet" type="text/css" href="../assets/styles.css"><script src="../assets/scripts.js" type="text/javascript"></script><!-- mathjax --><script type="text/x-mathjax-config">
    MathJax.Hub.Config({
      config: ["MMLorHTML.js"],
      jax: ["input/TeX","input/MathML","output/HTML-CSS","output/NativeMML", "output/PreviewHTML"],
      extensions: ["tex2jax.js","mml2jax.js","MathMenu.js","MathZoom.js", "fast-preview.js", "AssistiveMML.js", "a11y/accessibility-menu.js"],
      TeX: {
        extensions: ["AMSmath.js","AMSsymbols.js","noErrors.js","noUndefined.js"]
      },
      tex2jax: {
        inlineMath: [ ['$','$'], ['\\(', '\\)']],
        displayMath: [ ['$$','$$'], ['\\[', '\\]'] ],
        processEscapes: true
      }
    });
    </script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><!-- Responsive Favicon for The Carpentries --><link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png"><link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png"><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png"><link rel="manifest" href="../site.webmanifest"><link rel="mask-icon" href="../safari-pinned-tab.svg" color="#5bbad5"><meta name="msapplication-TileColor" content="#da532c"><meta name="theme-color" content="#ffffff"></head><body>
    <header id="top" class="navbar navbar-expand-md navbar-light bg-white top-nav incubator"><a class="visually-hidden-focusable skip-link" href="#main-content">Skip to main content</a>
  <div class="container-fluid top-nav-container">
    <div class="col-md-6">
      <div class="large-logo">
        <img alt="Carpentries Incubator" src="../assets/images/incubator-logo.svg"><abbr class="icon" title="This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team." style="text-decoration: unset">
          Â 
          <a href="https://cdh.carpentries.org/the-lesson-life-cycle.html#polishing-beta-stage" class="external-link alert-link" style="color: #383838">Beta
            <i aria-hidden="true" class="icon" data-feather="alert-circle" style="color: #001483; border-radius: 5px"></i>
          </a>
          <span class="visually-hidden">This lesson is in the beta phase, which means that it is ready for teaching by instructors outside of the original author team.</span>
        </abbr>
        
      </div>
    </div>
    <div class="selector-container ">
      
      
      <div class="dropdown">
        <button class="btn btn-secondary dropdown-toggle bordered-button" type="button" id="dropdownMenu1" data-bs-toggle="dropdown" aria-expanded="false">
          <i aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View <i data-feather="chevron-down"></i>
        </button>
        <ul class="dropdown-menu" aria-labelledby="dropdownMenu1"><li><button class="dropdown-item" type="button" onclick="window.location.href='../30-dplyr.html';">Learner View</button></li>
        </ul></div>
    </div>
  </div>
  <hr></header><nav class="navbar navbar-expand-xl navbar-light bg-white bottom-nav incubator" aria-label="Main Navigation"><div class="container-fluid nav-container">
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
      <span class="menu-title">Menu</span>
    </button>
    <div class="nav-logo">
      <img class="small-logo" alt="Carpentries Incubator" src="../assets/images/incubator-logo-sm.svg"></div>
    <div class="lesson-title-md">
      Introduction to data analysis with R and Bioconductor
    </div>
    <div class="search-icon-sm">
      <!-- TODO: do not show until we have search
      <i role="img" aria-label="search button" data-feather="search"></i>
      -->
    </div>
    <div class="desktop-nav">
      <ul class="navbar-nav me-auto mb-2 mb-lg-0"><li class="nav-item">
          <span class="lesson-title">
            Introduction to data analysis with R and Bioconductor
          </span>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/key-points.html">Key Points</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/instructor-notes.html">Instructor Notes</a>
        </li>
        <li class="nav-item">
          <a class="nav-link" href="../instructor/images.html">Extract All Images</a>
        </li>
        <li class="nav-item dropdown">
          <button class="nav-link dropdown-toggle" id="navbarDropdown" data-bs-toggle="dropdown" aria-expanded="false">
            More <i data-feather="chevron-down"></i>
          </button>
          <ul class="dropdown-menu" aria-labelledby="navbarDropdown"></ul></li>
      </ul></div>
    <form class="d-flex col-md-2 search-form">
      <fieldset disabled><input class="form-control me-2 searchbox" type="search" placeholder="Search" aria-label="Search"><button class="btn btn-outline-success tablet-search-button" type="submit">
          <i class="search-icon" data-feather="search" role="img" aria-label="search button"></i>
        </button>
      </fieldset></form>
  </div><!--/div.container-fluid -->
</nav><div class="col-md-12 mobile-title">
  Introduction to data analysis with R and Bioconductor
</div>

<aside class="col-md-12 lesson-progress"><div style="width: 43%" class="percentage">
    43%
  </div>
  <div class="progress incubator">
    <div class="progress-bar incubator" role="progressbar" style="width: 43%" aria-valuenow="43" aria-label="Lesson Progress" aria-valuemin="0" aria-valuemax="100">
    </div>
  </div>
</aside><div class="container">
      <div class="row">
        <!-- START: inst/pkgdown/templates/navbar.html -->
  <div id="sidebar-col" class="col-lg-4">
    <div id="sidebar" class="sidebar">
      <nav aria-labelledby="flush-headingEleven"><button role="button" aria-label="close menu" alt="close menu" aria-expanded="true" aria-controls="sidebar" class="collapse-toggle"><i class="search-icon" data-feather="x" role="img"></i></button>
        <div class="sidebar-inner">
          <div class="row mobile-row">
            <div class="col">
              <div class="sidenav-view-selector">
                <div class="accordion accordion-flush" id="accordionFlush9">
                  <div class="accordion-item">
                    <h2 class="accordion-header" id="flush-headingNine">
                      <button class="accordion-button collapsed" id="instructor" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseNine" aria-expanded="false" aria-controls="flush-collapseNine">
                        <i id="eye" aria-hidden="true" class="icon" data-feather="eye"></i> Instructor View
                      </button>
                    </h2>
                    <div id="flush-collapseNine" class="accordion-collapse collapse" aria-labelledby="flush-headingNine" data-bs-parent="#accordionFlush2">
                      <div class="accordion-body">
                        <a href="../30-dplyr.html">Learner View</a>
                      </div>
                    </div>
                  </div><!--/div.accordion-item-->
                </div><!--/div.accordion-flush-->
              </div><!--div.sidenav-view-selector -->
            </div><!--/div.col -->
      
            <hr></div><!--/div.mobile-row -->

          <div class="accordion accordion-flush" id="accordionFlush11">
            <div class="accordion-item">

              <button id="chapters" class="accordion-button show" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseEleven" aria-expanded="false" aria-controls="flush-collapseEleven">
                <h2 class="accordion-header chapters" id="flush-headingEleven">
                  EPISODES
                </h2>
              </button>
              <div id="flush-collapseEleven" class="accordion-collapse show collapse" aria-labelledby="flush-headingEleven" data-bs-parent="#accordionFlush11">

                <div class="accordion-body">
                  <div class="accordion accordion-flush" id="accordionFlush1">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading1">
        <a href="index.html">Summary and Schedule</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush2">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading2">
        <a href="10-data-organisation.html">1. Data organisation with spreadsheets</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush3">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading3">
        <a href="20-r-rstudio.html">2. R and RStudio</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush4">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading4">
        <a href="23-starting-with-r.html">3. Introduction to R</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush5">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading5">
        <a href="25-starting-with-data.html">4. Starting with data</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlushcurrent">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-headingcurrent">
      <button class="accordion-button" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapsecurrent" aria-expanded="true" aria-controls="flush-collapsecurrent">
        <span class="visually-hidden">Current Chapter</span>
        <span class="current-chapter">
        Manipulating and analysing data with dplyr
        </span>
      </button>
    </div><!--/div.accordion-header-->
        
    <div id="flush-collapsecurrent" class="accordion-collapse collapse show" aria-labelledby="flush-headingcurrent" data-bs-parent="#accordionFlushcurrent">
      <div class="accordion-body">
        <ul><li><a href="#data-manipulation-using-dplyr-and-tidyr">Data manipulation using <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong></a></li>
<li><a href="#loading-data-with-tidyverse">Loading data with tidyverse</a></li>
<li><a href="#selecting-columns-and-filtering-rows">Selecting columns and filtering rows</a></li>
<li><a href="#pipes">Pipes</a></li>
<li><a href="#mutate">Mutate</a></li>
<li><a href="#split-apply-combine-data-analysis">Split-apply-combine data analysis</a></li>
<li><a href="#reshaping-data">Reshaping data</a></li>
<li><a href="#joining-tables">Joining tables</a></li>
<li><a href="#exporting-data">Exporting data</a></li>
        </ul></div><!--/div.accordion-body-->
    </div><!--/div.accordion-collapse-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush7">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading7">
        <a href="40-visualization.html">6. Data visualization</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

<div class="accordion accordion-flush" id="accordionFlush8">
  <div class="accordion-item">
    <div class="accordion-header" id="flush-heading8">
        <a href="60-next-steps.html">7. Next steps</a>
    </div><!--/div.accordion-header-->
        
  </div><!--/div.accordion-item-->
</div><!--/div.accordion-flush-->

                </div>
              </div>
            </div>

            <hr class="half-width"><div class="accordion accordion-flush resources" id="accordionFlush12">
              <div class="accordion-item">
                <h2 class="accordion-header" id="flush-headingTwelve">
                  <button class="accordion-button collapsed" id="resources" type="button" data-bs-toggle="collapse" data-bs-target="#flush-collapseTwelve" aria-expanded="false" aria-controls="flush-collapseTwelve">
                    RESOURCES
                  </button>
                </h2>
                <div id="flush-collapseTwelve" class="accordion-collapse collapse" aria-labelledby="flush-headingTwelve" data-bs-parent="#accordionFlush12">
                  <div class="accordion-body">
                    <ul><li>
                        <a href="../instructor/key-points.html">Key Points</a>
                      </li>
                      <li>
                        <a href="../instructor/instructor-notes.html">Instructor Notes</a>
                      </li>
                      <li>
                        <a href="../instructor/images.html">Extract All Images</a>
                      </li>
                      
                    </ul></div>
                </div>
              </div>
            </div>
            <hr class="half-width resources"><a href="../instructor/aio.html">See all in one page</a>

            <hr class="d-none d-sm-block d-md-none"><div class="d-grid gap-1">
            
            </div>
          </div><!-- /div.accordion -->
        </div><!-- /div.sidebar-inner -->
      </nav></div><!-- /div.sidebar -->
  </div><!-- /div.sidebar-col -->
<!-- END:   inst/pkgdown/templates/navbar.html-->

        <!-- START: inst/pkgdown/templates/content-instructor.html -->
  <div class="col-xl-8 col-lg-12 primary-content">
    <nav class="lesson-content mx-md-4" aria-label="Previous and Next Chapter"><!-- content for small screens --><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Starting with data</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html" rel="next">Next: Data visualization... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <hr></nav><main id="main-content" class="main-content"><div class="container lesson-content">
        <h1>Manipulating and analysing data with dplyr</h1>
        <p> Last updated on 2023-04-03 | 
        
        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/30-dplyr.Rmd" class="external-link">Edit this page <i aria-hidden="true" data-feather="edit"></i></a></p>
        
        
        
        <p>Estimated time <i aria-hidden="true" data-feather="clock"></i> 150 minutes </p>
        
        <div class="text-end">
          <button role="button" aria-pressed="false" tabindex="0" id="expand-code" class="pull-right"> Expand All Solutions <i aria-hidden="true" data-feather="plus"></i></button>
        </div>

        

<div class="overview card">
<h2 class="card-header">Overview</h2>
<div class="row g-0">
<div class="col-md-4">
<div class="card-body">
<div class="inner">
<h3 class="card-title">Questions</h3>
<ul><li>Data analysis in R using the tidyverse meta-package</li>
</ul></div>
</div>
</div>
<div class="col-md-8">
<div class="card-body">
<div class="inner bordered">
<h3 class="card-title">Objectives</h3>
<ul><li>Describe the purpose of the <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong> packages.</li>
<li>Describe several of their functions that are extremely useful to
manipulate data.</li>
<li>Describe the concept of a wide and a long table format, and see how
to reshape a data frame from one format to the other one.</li>
</ul></div>
</div>
</div>
</div>
</div>
<section id="data-manipulation-using-dplyr-and-tidyr"><h2 class="section-heading">Data manipulation using <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong><a class="anchor" aria-label="anchor" href="#data-manipulation-using-dplyr-and-tidyr"></a>
</h2>
<hr class="half-width"><p>Bracket subsetting is handy, but it can be cumbersome and difficult
to read, especially for complicated operations.</p>
<p>Some packages can greatly facilitate our task when we manipulate
data. Packages in R are basically sets of additional functions that let
you do more stuff. The functions weâve been using so far, like
<code>str()</code> or <code>data.frame()</code>, come built into R;
Loading packages can give you access to other specific functions. Before
you use a package for the first time you need to install it on your
machine, and then you should import it in every subsequent R session
when you need it.</p>
<ul><li><p>The package <strong><code>dplyr</code></strong> provides powerful
tools for data manipulation tasks. It is built to work directly with
data frames, with many manipulation tasks optimised.</p></li>
<li><p>As we will see latter on, sometimes we want a data frame to be
reshaped to be able to do some specific analyses or for visualisation.
The package <strong><code>tidyr</code></strong> addresses this common
problem of reshaping data and provides tools for manipulating data in a
tidy way.</p></li>
</ul><p>To learn more about <strong><code>dplyr</code></strong> and
<strong><code>tidyr</code></strong> after the workshop, you may want to
check out this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-transformation.pdf" class="external-link">handy
data transformation with <strong><code>dplyr</code></strong>
cheatsheet</a> and this <a href="https://raw.githubusercontent.com/rstudio/cheatsheets/main/tidyr.pdf" class="external-link">one
about <strong><code>tidyr</code></strong></a>.</p>
<ul><li>The <strong><code>tidyverse</code></strong> package is an
âumbrella-packageâ that installs several useful packages for data
analysis which work well together, such as
<strong><code>tidyr</code></strong>,
<strong><code>dplyr</code></strong>,
<strong><code>ggplot2</code></strong>,
<strong><code>tibble</code></strong>, etc. These packages help us to
work and interact with the data. They allow us to do many things with
your data, such as subsetting, transforming, visualising, etc.</li>
</ul><p>If you did the set up, you should have already installed the
tidyverse package. Check to see if you have it by trying to load in from
the library:</p>
<div class="codewrapper sourceCode" id="cb1">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## load the tidyverse packages, incl. dplyr</span></span>
<span><span class="kw">library</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>If you got an error message
<code>there is no package called âtidyverseâ</code> then you have not
installed the package yet for this version of R. To install the
<strong><code>tidyverse</code></strong> package type:</p>
<div class="codewrapper sourceCode" id="cb2">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"tidyverse"</span><span class="op">)</span></span></code></pre>
</div>
<p>If you had to install the <strong><code>tidyverse</code></strong>
package, do not forget to load it in this R session by using the
<code>library()</code> command above!</p>
</section><section id="loading-data-with-tidyverse"><h2 class="section-heading">Loading data with tidyverse<a class="anchor" aria-label="anchor" href="#loading-data-with-tidyverse"></a>
</h2>
<hr class="half-width"><p>Instead of <code>read.csv()</code>, we will read in our data using
the <code>read_csv()</code> function (notice the <code>_</code> instead
of the <code>.</code>), from the tidyverse package
<strong><code>readr</code></strong>.</p>
<div class="codewrapper sourceCode" id="cb3">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/rnaseq.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">WARNING<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="warning" tabindex="0"><code>Warning: One or more parsing issues, call `problems()` on your data frame for details,
e.g.:
  dat &lt;- vroom(...)
  problems(dat)</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb5">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## view the data</span></span>
<span><span class="va">rna</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 1,129 Ã 1
   `&lt;!DOCTYPE html&gt;`                                                            
   &lt;chr&gt;                                                                        
 1 "&lt;html lang=\"en\" data-color-mode=\"auto\" data-light-theme=\"light\" data-â¦
 2 "&lt;head&gt;"                                                                     
 3 "&lt;meta charset=\"utf-8\"&gt;"                                                   
 4 "&lt;link rel=\"dns-prefetch\" href=\"https://github.githubassets.com\"&gt;"       
 5 "&lt;link rel=\"dns-prefetch\" href=\"https://avatars.githubusercontent.com\"&gt;" 
 6 "&lt;link rel=\"dns-prefetch\" href=\"https://github-cloud.s3.amazonaws.com\"&gt;" 
 7 "&lt;link rel=\"dns-prefetch\" href=\"https://user-images.githubusercontent.comâ¦
 8 "&lt;link rel=\"preconnect\" href=\"https://github.githubassets.com\" crossorigâ¦
 9 "&lt;link rel=\"preconnect\" href=\"https://avatars.githubusercontent.com\"&gt;"   
10 "&lt;link crossorigin=\"anonymous\" media=\"all\" rel=\"stylesheet\" href=\"httâ¦
# â¦ with 1,119 more rows</code></pre>
</div>
<p>Notice that the class of the data is now referred to as a
âtibbleâ.</p>
<p>Tibbles tweak some of the behaviors of the data frame objects we
introduced in the previously. The data structure is very similar to a
data frame. For our purposes the only differences are that:</p>
<ol style="list-style-type: decimal"><li><p>It displays the data type of each column under its name. Note
that &lt;<code>dbl</code>&gt; is a data type defined to hold numeric
values with decimal points.</p></li>
<li><p>It only prints the first few rows of data and only as many
columns as fit on one screen.</p></li>
</ol><p>We are now going to learn some of the most common
<strong><code>dplyr</code></strong> functions:</p>
<ul><li>
<code>select()</code>: subset columns</li>
<li>
<code>filter()</code>: subset rows on conditions</li>
<li>
<code>mutate()</code>: create new columns by using information from
other columns</li>
<li>
<code>group_by()</code> and <code>summarise()</code>: create summary
statistics on grouped data</li>
<li>
<code>arrange()</code>: sort results</li>
<li>
<code>count()</code>: count discrete values</li>
</ul></section><section id="selecting-columns-and-filtering-rows"><h2 class="section-heading">Selecting columns and filtering rows<a class="anchor" aria-label="anchor" href="#selecting-columns-and-filtering-rows"></a>
</h2>
<hr class="half-width"><p>To select columns of a data frame, use <code>select()</code>. The
first argument to this function is the data frame (<code>rna</code>),
and the subsequent arguments are the columns to keep.</p>
<div class="codewrapper sourceCode" id="cb7">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<p>To select all columns <em>except</em> certain ones, put a â-â in
front of the variable to exclude it.</p>
<div class="codewrapper sourceCode" id="cb9">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="op">-</span><span class="va">tissue</span>, <span class="op">-</span><span class="va">organism</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `tissue` doesn't exist.</code></pre>
</div>
<p>This will select all the variables in <code>rna</code> except
<code>tissue</code> and <code>organism</code>.</p>
<p>To choose rows based on a specific criteria, use
<code>filter()</code>:</p>
<div class="codewrapper sourceCode" id="cb11">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb13">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span> <span class="op">&amp;</span> <span class="va">infection</span> <span class="op">==</span> <span class="st">"NonInfected"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male" &amp; infection == "NonInfected"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<p>Now letâs imagine we are interested in the human homologs of the
mouse genes analysed in this dataset. This information can be found in
the last column of the <code>rna</code> tibble, named
<code>hsapiens_homolog_associated_gene_name</code>. To visualise it
easily, we will create a new table containing just the 2 columns
<code>gene</code> and
<code>hsapiens_homolog_associated_gene_name</code>.</p>
<div class="codewrapper sourceCode" id="cb15">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna</span>, <span class="va">gene</span>, <span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb17">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">genes</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'genes' not found</code></pre>
</div>
<p>Some mouse genes have no human homologs. These can be retrieved using
<code>filter()</code> and the <code>is.na()</code> function, that
determines whether something is an <code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb19">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in filter(genes, is.na(hsapiens_homolog_associated_gene_name)): object 'genes' not found</code></pre>
</div>
<p>If we want to keep only mouse genes that have a human homolog, we can
insert a â!â symbol that negates the result, so weâre asking for every
row where hsapiens_homolog_associated_gene_name <em>is not</em> an
<code>NA</code>.</p>
<div class="codewrapper sourceCode" id="cb21">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">filter</span><span class="op">(</span><span class="va">genes</span>, <span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">hsapiens_homolog_associated_gene_name</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in filter(genes, !is.na(hsapiens_homolog_associated_gene_name)): object 'genes' not found</code></pre>
</div>
</section><section id="pipes"><h2 class="section-heading">Pipes<a class="anchor" aria-label="anchor" href="#pipes"></a>
</h2>
<hr class="half-width"><p>What if you want to select and filter at the same time? There are
three ways to do this: use intermediate steps, nested functions, or
pipes.</p>
<p>With intermediate steps, you create a temporary data frame and use
that as input to the next function, like this:</p>
<div class="codewrapper sourceCode" id="cb23">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna2</span> <span class="op">&lt;-</span> <span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb25">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="va">rna2</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in select(rna2, gene, sample, tissue, expression): object 'rna2' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb27">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna3' not found</code></pre>
</div>
<p>This is readable, but can clutter up your workspace with lots of
intermediate objects that you have to name individually. With multiple
steps, that can be hard to keep track of.</p>
<p>You can also nest functions (i.e.Â one function inside of another),
like this:</p>
<div class="codewrapper sourceCode" id="cb29">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="fu">select</span><span class="op">(</span><span class="fu">filter</span><span class="op">(</span><span class="va">rna</span>, <span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span>, <span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb31">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna3' not found</code></pre>
</div>
<p>This is handy, but can be difficult to read if too many functions are
nested, as R evaluates the expression from the inside out (in this case,
filtering, then selecting).</p>
<p>The last option, <em>pipes</em>, are a recent addition to R. Pipes
let you take the output of one function and send it directly to the
next, which is useful when you need to do many things to the same
dataset.</p>
<p>Pipes in R look like <code>%&gt;%</code> (made available via the
<strong><code>magrittr</code></strong> package) or <code>|&gt;</code>
(through base R). If you use RStudio, you can type the pipe with
<kbd>Ctrl</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you have a PC or
<kbd>Cmd</kbd> + <kbd>Shift</kbd> + <kbd>M</kbd> if you have a Mac.</p>
<p>In the above code, we use the pipe to send the <code>rna</code>
dataset first through <code>filter()</code> to keep rows where
<code>sex</code> is Male, then through <code>select()</code> to keep
only the <code>gene</code>, <code>sample</code>, <code>tissue</code>,
and <code>expression</code>columns.</p>
<p>The pipe <code>%&gt;%</code> takes the object on its left and passes
it directly as the first argument to the function on its right, we donât
need to explicitly include the data frame as an argument to the
<code>filter()</code> and <code>select()</code> functions any more.</p>
<div class="codewrapper sourceCode" id="cb33">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<p>Some may find it helpful to read the pipe like the word âthenâ. For
instance, in the above example, we took the data frame <code>rna</code>,
<em>then</em> we <code>filter</code>ed for rows with
<code>sex == "Male"</code>, <em>then</em> we <code>select</code>ed
columns <code>gene</code>, <code>sample</code>, <code>tissue</code>, and
<code>expression</code>.</p>
<p>The <strong><code>dplyr</code></strong> functions by themselves are
somewhat simple, but by combining them into linear workflows with the
pipe, we can accomplish more complex manipulations of data frames.</p>
<p>If we want to create a new object with this smaller version of the
data, we can assign it a new name:</p>
<div class="codewrapper sourceCode" id="cb35">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sex</span> <span class="op">==</span> <span class="st">"Male"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">tissue</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sex == "Male"`.
Caused by error:
! object 'sex' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb37">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna3</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna3' not found</code></pre>
</div>
<div id="challenge1" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge" class="callout-inner">
<h3 class="callout-title">Challenge:<a class="anchor" aria-label="anchor" href="#challenge1"></a>
</h3>
<div class="callout-content">
<p>Using pipes, subset the <code>rna</code> data to keep observations in
female mice at time 0, where the gene has an expression higher than
50000, and retain only the columns <code>gene</code>,
<code>sample</code>, <code>time</code>, <code>expression</code> and
<code>age</code>.</p>
</div>
</div>
</div>
<div id="accordionSolution1" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution1" aria-expanded="false" aria-controls="collapseSolution1">
  <h4 class="accordion-header" id="headingSolution1">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution1" class="accordion-collapse collapse" data-bs-parent="#accordionSolution1" aria-labelledby="headingSolution1">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb39">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">50000</span>,</span>
<span>         <span class="va">sex</span> <span class="op">==</span> <span class="st">"Female"</span>,</span>
<span>         <span class="va">time</span> <span class="op">==</span> <span class="fl">0</span> <span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">time</span>, <span class="va">expression</span>, <span class="va">age</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `expression &gt; 50000`.
Caused by error in `expression &gt; 50000`:
! comparison (&gt;) is possible only for atomic and list types</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="mutate"><h2 class="section-heading">Mutate<a class="anchor" aria-label="anchor" href="#mutate"></a>
</h2>
<hr class="half-width"><p>Frequently youâll want to create new columns based on the values of
existing columns, for example to do unit conversions, or to find the
ratio of values in two columns. For this weâll use
<code>mutate()</code>.</p>
<p>To create a new column of time in hours:</p>
<div class="codewrapper sourceCode" id="cb41">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `mutate()`:
â¹ In argument: `time_hours = time * 24`.
Caused by error in `time * 24`:
! non-numeric argument to binary operator</code></pre>
</div>
<p>You can also create a second new column based on the first new column
within the same call of <code>mutate()</code>:</p>
<div class="codewrapper sourceCode" id="cb43">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_hours <span class="op">=</span> <span class="va">time</span> <span class="op">*</span> <span class="fl">24</span>,</span>
<span>         time_mn <span class="op">=</span> <span class="va">time_hours</span> <span class="op">*</span> <span class="fl">60</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">time</span>, <span class="va">time_hours</span>, <span class="va">time_mn</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `mutate()`:
â¹ In argument: `time_hours = time * 24`.
Caused by error in `time * 24`:
! non-numeric argument to binary operator</code></pre>
</div>
<div id="challenge2" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-1" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge2"></a>
</h3>
<div class="callout-content">
<p>Create a new data frame from the <code>rna</code> data that meets the
following criteria: contains only the <code>gene</code>,
<code>chromosome_name</code>, <code>phenotype_description</code>,
<code>sample</code>, and <code>expression</code> columns. The expression
values should be log-transformed. This data frame must only contain
genes located on sex chromosomes, associated with a
phenotype_description, and with a log expression higher than 5.</p>
<p><strong>Hint</strong>: think about how the commands should be ordered
to produce this data frame!</p>
</div>
</div>
</div>
<div id="accordionSolution2" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution2" aria-expanded="false" aria-controls="collapseSolution2">
  <h4 class="accordion-header" id="headingSolution2">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution2" class="accordion-collapse collapse" data-bs-parent="#accordionSolution2" aria-labelledby="headingSolution2">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb45">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>expression <span class="op">=</span> <span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">chromosome_name</span>, <span class="va">phenotype_description</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="op">!</span><span class="fu">is.na</span><span class="op">(</span><span class="va">phenotype_description</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">expression</span> <span class="op">&gt;</span> <span class="fl">5</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `mutate()`:
â¹ In argument: `expression = log(expression)`.
Caused by error in `log()`:
! non-numeric argument to mathematical function</code></pre>
</div>
</div>
</div>
</div>
</div>
</section><section id="split-apply-combine-data-analysis"><h2 class="section-heading">Split-apply-combine data analysis<a class="anchor" aria-label="anchor" href="#split-apply-combine-data-analysis"></a>
</h2>
<hr class="half-width"><p>Many data analysis tasks can be approached using the
<em>split-apply-combine</em> paradigm: split the data into groups, apply
some analysis to each group, and then combine the results.
<strong><code>dplyr</code></strong> makes this very easy through the use
of the <code>group_by()</code> function.</p>
<div class="codewrapper sourceCode" id="cb47">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `gene` is not found.</code></pre>
</div>
<p>The <code>group_by()</code> function doesnât perform any data
processing, it groups the data into subsets: in the example above, our
initial <code>tibble</code> of 1129 observations is split into 0 groups
based on the <code>gene</code> variable.</p>
<p>We could similarly decide to group the tibble by the samples:</p>
<div class="codewrapper sourceCode" id="cb49">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `sample` is not found.</code></pre>
</div>
<p>Here our initial <code>tibble</code> of 1129 observations is split
into 0 groups based on the <code>sample</code> variable.</p>
<p>Once the data has been grouped, subsequent operations will be applied
on each group independently.</p>
<div class="section level3">
<h3 id="the-summarise-function">The <code>summarise()</code> function<a class="anchor" aria-label="anchor" href="#the-summarise-function"></a></h3>
<p><code>group_by()</code> is often used together with
<code>summarise()</code>, which collapses each group into a single-row
summary of that group.</p>
<p><code>group_by()</code> takes as arguments the column names that
contain the <strong>categorical</strong> variables for which you want to
calculate the summary statistics. So to compute the mean
<code>expression</code> by gene:</p>
<div class="codewrapper sourceCode" id="cb51">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `gene` is not found.</code></pre>
</div>
<p>We could also want to calculate the mean expression levels of all
genes in each sample:</p>
<div class="codewrapper sourceCode" id="cb53">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `sample` is not found.</code></pre>
</div>
<p>But we can can also group by multiple columns:</p>
<div class="codewrapper sourceCode" id="cb55">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<p>Once the data is grouped, you can also summarise multiple variables
at the same time (and not necessarily on the same variable). For
instance, we could add a column indicating the median
<code>expression</code> by gene and by condition:</p>
<div class="codewrapper sourceCode" id="cb57">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span>,</span>
<span>            median_expression <span class="op">=</span> <span class="fu">median</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<div id="challenge3" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-2" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge3"></a>
</h3>
<div class="callout-content">
<p>Calculate the mean expression level of gene âDok3â by timepoints.</p>
</div>
</div>
</div>
<div id="accordionSolution3" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution3" aria-expanded="false" aria-controls="collapseSolution3">
  <h4 class="accordion-header" id="headingSolution3">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution3" class="accordion-collapse collapse" data-bs-parent="#accordionSolution3" aria-labelledby="headingSolution3">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb59">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">gene</span> <span class="op">==</span> <span class="st">"Dok3"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `gene == "Dok3"`.
Caused by error:
! object 'gene' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
<div class="section level3">
<h3 id="counting">Counting<a class="anchor" aria-label="anchor" href="#counting"></a></h3>
<p>When working with data, we often want to know the number of
observations found for each factor or combination of factors. For this
task, <strong><code>dplyr</code></strong> provides <code>count()</code>.
For example, if we wanted to count the number of rows of data for each
infected and non-infected samples, we would do:</p>
<div class="codewrapper sourceCode" id="cb61">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
â Column `infection` is not found.</code></pre>
</div>
<p>The <code>count()</code> function is shorthand for something weâve
already seen: grouping by a variable, and summarising it by counting the
number of observations in that group. In other words,
<code>rna %&gt;% count(infection)</code> is equivalent to:</p>
<div class="codewrapper sourceCode" id="cb63">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `infection` is not found.</code></pre>
</div>
<p>The previous example shows the use of <code>count()</code> to count
the number of rows/observations for <em>one</em> factor (i.e.,
<code>infection</code>). If we wanted to count a <em>combination of
factors</em>, such as <code>infection</code> and <code>time</code>, we
would specify the first and the second factor as the arguments of
<code>count()</code>:</p>
<div class="codewrapper sourceCode" id="cb65">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<p>which is equivalent to this:</p>
<div class="codewrapper sourceCode" id="cb67">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>n <span class="op">=</span> <span class="fu">n</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<p>It is sometimes useful to sort the result to facilitate the
comparisons. We can use <code>arrange()</code> to sort the table. For
instance, we might want to arrange the table above by time:</p>
<div class="codewrapper sourceCode" id="cb69">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">time</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<p>or by counts:</p>
<div class="codewrapper sourceCode" id="cb71">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">n</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<p>To sort in descending order, we need to add the <code>desc()</code>
function:</p>
<div class="codewrapper sourceCode" id="cb73">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">infection</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
Column `infection` is not found.
Column `time` is not found.</code></pre>
</div>
<div id="challenge4" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-3" class="callout-inner">
<h3 class="callout-title">Challenge<a class="anchor" aria-label="anchor" href="#challenge4"></a>
</h3>
<div class="callout-content">
<ol style="list-style-type: decimal"><li>How many genes were analysed in each sample?</li>
<li>Use <code>group_by()</code> and <code>summarise()</code> to evaluate
the sequencing depth (the sum of all counts) in each sample. Which
sample has the highest sequencing depth?</li>
<li>Pick one sample and evaluate the number of genes by biotype.</li>
<li>Identify genes associated with the âabnormal DNA methylationâ
phenotype description, and calculate their mean expression (in log) at
time 0, time 4 and time 8.</li>
</ol></div>
</div>
</div>
<div id="accordionSolution4" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution4" aria-expanded="false" aria-controls="collapseSolution4">
  <h4 class="accordion-header" id="headingSolution4">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution4" class="accordion-collapse collapse" data-bs-parent="#accordionSolution4" aria-labelledby="headingSolution4">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb75">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 1.</span></span>
<span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `count()`:
! Must group by variables found in `.data`.
â Column `sample` is not found.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb77">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 2.</span></span>
<span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>seq_depth <span class="op">=</span> <span class="fu">sum</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">seq_depth</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
â Column `sample` is not found.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb79">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 3.</span></span>
<span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">sample</span> <span class="op">==</span> <span class="st">"GSM2545336"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">count</span><span class="op">(</span><span class="va">gene_biotype</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="fu">desc</span><span class="op">(</span><span class="va">n</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `sample == "GSM2545336"`.
Caused by error in `sample == "GSM2545336"`:
! comparison (==) is possible only for atomic and list types</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb81">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">## 4.</span></span>
<span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">phenotype_description</span> <span class="op">==</span> <span class="st">"abnormal DNA methylation"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_expression <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="fu">log</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `phenotype_description == "abnormal DNA methylation"`.
Caused by error:
! object 'phenotype_description' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="reshaping-data"><h2 class="section-heading">Reshaping data<a class="anchor" aria-label="anchor" href="#reshaping-data"></a>
</h2>
<hr class="half-width"><p>In the <code>rna</code> tibble, the rows contain expression values
(the unit) that are associated with a combination of 2 other variables:
<code>gene</code> and <code>sample</code>.</p>
<p>All the other columns correspond to variables describing either the
sample (organism, age, sex, â¦) or the gene (gene_biotype, ENTREZ_ID,
product, â¦). The variables that donât change with genes or with samples
will have the same value in all the rows.</p>
<div class="codewrapper sourceCode" id="cb83">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">arrange</span><span class="op">(</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `arrange()`:
â¹ In argument: `..1 = gene`.
Caused by error:
! object 'gene' not found</code></pre>
</div>
<p>This structure is called a <code>long-format</code>, as one column
contains all the values, and other column(s) list(s) the context of the
value.</p>
<p>In certain cases, the <code>long-format</code> is not really
âhuman-readableâ, and another format, a <code>wide-format</code> is
preferred, as a more compact way of representing the data. This is
typically the case with gene expression values that scientists are used
to look as matrices, were rows represent genes and columns represent
samples.</p>
<p>In this format, it would therefore become straightforward to explore
the relationship between the gene expression levels within, and between,
the samples.</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<p>To convert the gene expression values from <code>rna</code> into a
wide-format, we need to create a new table where the values of the
<code>sample</code> column would become the names of column
variables.</p>
<p>The key point here is that we are still following a tidy data
structure, but we have <strong>reshaped</strong> the data according to
the observations of interest: expression levels per gene instead of
recording them per gene and per sample.</p>
<p>The opposite transformation would be to transform column names into
values of a new variable.</p>
<p>We can do both these of transformations with two <code>tidyr</code>
functions, <code>pivot_longer()</code> and <code>pivot_wider()</code>
(see <a href="https://tidyr.tidyverse.org/dev/articles/pivot.html" class="external-link">here</a> for
details).</p>
<div class="section level3">
<h3 id="pivoting-the-data-into-a-wider-format">Pivoting the data into a wider format<a class="anchor" aria-label="anchor" href="#pivoting-the-data-into-a-wider-format"></a></h3>
<p>Letâs select the first 3 columns of <code>rna</code> and use
<code>pivot_wider()</code> to transform the data into a wide-format.</p>
<div class="codewrapper sourceCode" id="cb86">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_exp</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb88">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_exp</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_exp' not found</code></pre>
</div>
<p><code>pivot_wider</code> takes three main arguments:</p>
<ol style="list-style-type: decimal"><li>the data to be transformed;</li>
<li>the <code>names_from</code> : the column whose values will become
new column names;</li>
<li>the <code>values_from</code>: the column whose values will fill the
new columns.</li>
</ol><figure style="text-align: center"><img src="../fig/pivot_wider.png" alt="Wide pivot of the `rna` data." class="figure mx-auto d-block"><figcaption>
Wide pivot of the <code>rna</code> data.
</figcaption></figure><div class="codewrapper sourceCode" id="cb90">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">&lt;-</span> <span class="va">rna_exp</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_wider(., names_from = sample, values_from = expression): object 'rna_exp' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb92">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_wide' not found</code></pre>
</div>
<p>Note that by default, the <code>pivot_wider()</code> function will
add <code>NA</code> for missing values.</p>
<p>Letâs imagine that for some reason, we had some missing expression
values for some genes in certain samples. In the following fictive
example, the gene Cyp2d22 has only one expression value, in GSM2545338
sample.</p>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb95">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_with_missing_values' not found</code></pre>
</div>
<p>By default, the <code>pivot_wider()</code> function will add
<code>NA</code> for missing values. This can be parameterised with the
<code>values_fill</code> argument of the <code>pivot_wider()</code>
function.</p>
<div class="codewrapper sourceCode" id="cb97">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_wider(., names_from = sample, values_from = expression): object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb99">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span>,</span>
<span>              values_fill <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_wider(., names_from = sample, values_from = expression, : object 'rna_with_missing_values' not found</code></pre>
</div>
</div>
<div class="section level3">
<h3 id="pivoting-data-into-a-longer-format">Pivoting data into a longer format<a class="anchor" aria-label="anchor" href="#pivoting-data-into-a-longer-format"></a></h3>
<p>In the opposite situation we are using the column names and turning
them into a pair of new variables. One variable represents the column
names as values, and the other variable contains the values previously
associated with the column names.</p>
<p><code>pivot_longer()</code> takes four main arguments:</p>
<ol style="list-style-type: decimal"><li>the data to be transformed;</li>
<li>the <code>names_to</code>: the new column name we wish to create and
populate with the current column names;</li>
<li>the <code>values_to</code>: the new column name we wish to create
and populate with current values;</li>
<li>the names of the columns to be used to populate the
<code>names_to</code> and <code>values_to</code> variables (or to
drop).</li>
</ol><figure style="text-align: center"><img src="../fig/pivot_longer.png" alt="Long pivot of the `rna` data." class="figure mx-auto d-block"><figcaption>
Long pivot of the <code>rna</code> data.
</figcaption></figure><p>To recreate <code>rna_long</code> from <code>rna_wide</code> we would
create a key called <code>sample</code> and value called
<code>expression</code> and use all columns except <code>gene</code> for
the key variable. Here we drop <code>gene</code> column with a minus
sign.</p>
<p>Notice how the new variable names are to be quoted here.</p>
<div class="codewrapper sourceCode" id="cb101">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_long</span> <span class="op">&lt;-</span> <span class="va">rna_wide</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "sample", values_to = "expression", : object 'rna_wide' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb103">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_long</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_long' not found</code></pre>
</div>
<p>We could also have used a specification for what columns to include.
This can be useful if you have a large number of identifying columns,
and itâs easier to specify what to gather than what to leave alone. Here
the <code>starts_with()</code> function can help to retrieve sample
names without having to list them all! Another possibility would be to
use the <code>:</code> operator!</p>
<div class="codewrapper sourceCode" id="cb105">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 cols <span class="op">=</span> <span class="fu">starts_with</span><span class="op">(</span><span class="st">"GSM"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "sample", values_to = "expression", : object 'rna_wide' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb107">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_wide</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="va">GSM2545336</span><span class="op">:</span><span class="va">GSM2545380</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "sample", values_to = "expression", : object 'rna_wide' not found</code></pre>
</div>
<p>Note that if we had missing values in the wide-format, the
<code>NA</code> would be included in the new long format.</p>
<p>Remember our previous fictive tibble containing missing values:</p>
<div class="codewrapper sourceCode" id="cb109">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_with_missing_values</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb111">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">&lt;-</span> <span class="va">rna_with_missing_values</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sample</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_wider(., names_from = sample, values_from = expression): object 'rna_with_missing_values' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb113">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'wide_with_NA' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb115">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">wide_with_NA</span> <span class="op">%&gt;%</span></span>
<span>    <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"sample"</span>,</span>
<span>                 values_to <span class="op">=</span> <span class="st">"expression"</span>,</span>
<span>                 <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "sample", values_to = "expression", : object 'wide_with_NA' not found</code></pre>
</div>
<p>Pivoting to wider and longer formats can be a useful way to balance
out a dataset so every replicate has the same composition.</p>
<div id="challenge5" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question" class="callout-inner">
<h3 class="callout-title">Question<a class="anchor" aria-label="anchor" href="#challenge5"></a>
</h3>
<div class="callout-content">
<p>Starting from the rna table, use the <code>pivot_wider()</code>
function to create a wide-format table giving the gene expression levels
in each mouse. Then use the <code>pivot_longer()</code> function to
restore a long-format table.</p>
</div>
</div>
</div>
<div id="accordionSolution5" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution5" aria-expanded="false" aria-controls="collapseSolution5">
  <h4 class="accordion-header" id="headingSolution5">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution5" class="accordion-collapse collapse" data-bs-parent="#accordionSolution5" aria-labelledby="headingSolution5">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb117">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">mouse</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">mouse</span>, values_from <span class="op">=</span> <span class="va">expression</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb119">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna1' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb121">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna1</span> <span class="op">%&gt;%</span></span>
<span><span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"mouse_id"</span>, values_to <span class="op">=</span> <span class="st">"counts"</span>, <span class="op">-</span><span class="va">gene</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "mouse_id", values_to = "counts", : object 'rna1' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge6" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-1" class="callout-inner">
<h3 class="callout-title">Question<a class="anchor" aria-label="anchor" href="#challenge6"></a>
</h3>
<div class="callout-content">
<p>Subset genes located on X and Y chromosomes from the <code>rna</code>
data frame and spread the data frame with <code>sex</code> as columns,
<code>chromosome_name</code> as rows, and the mean expression of genes
located in each chromosome as the values, as in the following
tibble:</p>
<figure><img src="../fig/Exercise_pivot_W.png" style="display: block; margin: auto;" class="figure mx-auto d-block"></figure><p>You will need to summarise before reshaping!</p>
</div>
</div>
</div>
<div id="accordionSolution6" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution6" aria-expanded="false" aria-controls="collapseSolution6">
  <h4 class="accordion-header" id="headingSolution6">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution6" class="accordion-collapse collapse" data-bs-parent="#accordionSolution6" aria-labelledby="headingSolution6">
<div class="accordion-body">
<p>Letâs first calculate the mean expression level of X and Y linked
genes from male and female samplesâ¦</p>
<div class="codewrapper sourceCode" id="cb123">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `chromosome_name == "Y" | chromosome_name == "X"`.
Caused by error:
! object 'chromosome_name' not found</code></pre>
</div>
<p>And pivot the table to wide format</p>
<div class="codewrapper sourceCode" id="cb125">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">filter</span><span class="op">(</span><span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"Y"</span> <span class="op">|</span> <span class="va">chromosome_name</span> <span class="op">==</span> <span class="st">"X"</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">sex</span>, <span class="va">chromosome_name</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">sex</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `filter()`:
â¹ In argument: `chromosome_name == "Y" | chromosome_name == "X"`.
Caused by error:
! object 'chromosome_name' not found</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb127">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_1' not found</code></pre>
</div>
<p>Now take that data frame and transform it with
<code>pivot_longer()</code> so each row is a unique
<code>chromosome_name</code> by <code>gender</code> combination.</p>
<div class="codewrapper sourceCode" id="cb129">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_1</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"gender"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"mean"</span>,</span>
<span>               <span class="op">-</span><span class="va">chromosome_name</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in pivot_longer(., names_to = "gender", values_to = "mean", -chromosome_name): object 'rna_1' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge7" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-2" class="callout-inner">
<h3 class="callout-title">Question<a class="anchor" aria-label="anchor" href="#challenge7"></a>
</h3>
<div class="callout-content">
<p>Use the <code>rna</code> dataset to create an expression matrix where
each row represents the mean expression levels of genes and columns
represent the different timepoints.</p>
</div>
</div>
</div>
<div id="accordionSolution7" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution7" aria-expanded="false" aria-controls="collapseSolution7">
  <h4 class="accordion-header" id="headingSolution7">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution7" class="accordion-collapse collapse" data-bs-parent="#accordionSolution7" aria-labelledby="headingSolution7">
<div class="accordion-body">
<p>Letâs first calculate the mean expression by gene and by time</p>
<div class="codewrapper sourceCode" id="cb131">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `time` is not found.</code></pre>
</div>
<p>before using the pivot_wider() function</p>
<div class="codewrapper sourceCode" id="cb133">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `time` is not found.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb135">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_time' not found</code></pre>
</div>
<p>Notice that this generates a tibble with some column names starting
by a number. If we wanted to select the column corresponding to the
timepoints, we could not use the column names directlyâ¦ What happens
when we select the column 4?</p>
<div class="codewrapper sourceCode" id="cb137">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="fl">4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `time` is not found.</code></pre>
</div>
<p>To select the timepoint 4, we would have to quote the column name,
with backticks â`â</p>
<div class="codewrapper sourceCode" id="cb139">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `time` is not found.</code></pre>
</div>
<p>Another possibility would be to rename the column, choosing a name
that doesnât start by a number :</p>
<div class="codewrapper sourceCode" id="cb141">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">group_by</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">summarise</span><span class="op">(</span>mean_exp <span class="op">=</span> <span class="fu">mean</span><span class="op">(</span><span class="va">expression</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_wider</span><span class="op">(</span>names_from <span class="op">=</span> <span class="va">time</span>,</span>
<span>              values_from <span class="op">=</span> <span class="va">mean_exp</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">rename</span><span class="op">(</span><span class="st">"time0"</span> <span class="op">=</span> <span class="va">`0`</span>, <span class="st">"time4"</span> <span class="op">=</span> <span class="va">`4`</span>, <span class="st">"time8"</span> <span class="op">=</span> <span class="va">`8`</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">time4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `group_by()`:
! Must group by variables found in `.data`.
Column `gene` is not found.
Column `time` is not found.</code></pre>
</div>
</div>
</div>
</div>
</div>
<div id="challenge8" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="question-3" class="callout-inner">
<h3 class="callout-title">Question<a class="anchor" aria-label="anchor" href="#challenge8"></a>
</h3>
<div class="callout-content">
<p>Use the previous data frame containing mean expression levels per
timepoint and create a new column containing fold-changes between
timepoint 8 and timepoint 0, and fold-changes between timepoint 8 and
timepoint 4. Convert this table into a long-format table gathering the
fold-changes calculated.</p>
</div>
</div>
</div>
<div id="accordionSolution8" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution8" aria-expanded="false" aria-controls="collapseSolution8">
  <h4 class="accordion-header" id="headingSolution8">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution8" class="accordion-collapse collapse" data-bs-parent="#accordionSolution8" aria-labelledby="headingSolution8">
<div class="accordion-body">
<p>Starting from the rna_time tibble:</p>
<div class="codewrapper sourceCode" id="cb143">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_time' not found</code></pre>
</div>
<p>Calculate fold-changes:</p>
<div class="codewrapper sourceCode" id="cb145">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in mutate(., time_8_vs_0 = `8`/`0`, time_8_vs_4 = `8`/`4`): object 'rna_time' not found</code></pre>
</div>
<p>And use the pivot_longer() function:</p>
<div class="codewrapper sourceCode" id="cb147">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_time</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">mutate</span><span class="op">(</span>time_8_vs_0 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`0`</span>, time_8_vs_4 <span class="op">=</span> <span class="va">`8`</span> <span class="op">/</span> <span class="va">`4`</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>  <span class="fu">pivot_longer</span><span class="op">(</span>names_to <span class="op">=</span> <span class="st">"comparisons"</span>,</span>
<span>               values_to <span class="op">=</span> <span class="st">"Fold_changes"</span>,</span>
<span>               <span class="va">time_8_vs_0</span><span class="op">:</span><span class="va">time_8_vs_4</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in mutate(., time_8_vs_0 = `8`/`0`, time_8_vs_4 = `8`/`4`): object 'rna_time' not found</code></pre>
</div>
</div>
</div>
</div>
</div>
</div>
</section><section id="joining-tables"><h2 class="section-heading">Joining tables<a class="anchor" aria-label="anchor" href="#joining-tables"></a>
</h2>
<hr class="half-width"><p>In many real life situations, data are spread across multiple tables.
Usually this occurs because different types of information are collected
from different sources.</p>
<p>It may be desirable for some analyses to combine data from two or
more tables into a single data frame based on a column that would be
common to all the tables.</p>
<p>The <code>dplyr</code> package provides a set of join functions for
combining two data frames based on matches within specified columns.
Here, we provide a short introduction to joins. For further reading,
please refer to the chapter about <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-join.html" class="external-link">table
joins</a>. The <a href="https://github.com/rstudio/cheatsheets/raw/master/data-transformation.pdf" class="external-link">Data
Transformation Cheat Sheet</a> also provides a short overview on table
joins.</p>
<p>We are going to illustrate join using a small table,
<code>rna_mini</code> that we will create by subsetting the original
<code>rna</code> table, keeping only 3 columns and 10 lines.</p>
<div class="codewrapper sourceCode" id="cb149">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_mini</span> <span class="op">&lt;-</span> <span class="va">rna</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu">select</span><span class="op">(</span><span class="va">gene</span>, <span class="va">sample</span>, <span class="va">expression</span><span class="op">)</span> <span class="op">%&gt;%</span></span>
<span>   <span class="fu">head</span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in `select()`:
! Can't subset columns that don't exist.
â Column `gene` doesn't exist.</code></pre>
</div>
<div class="codewrapper sourceCode" id="cb151">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">rna_mini</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in eval(expr, envir, enclos): object 'rna_mini' not found</code></pre>
</div>
<p>The second table, <code>annot1</code>, contains 2 columns, gene and
gene_description. You can either <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv" class="external-link">download
annot1.csv</a> by clicking on the link and then moving it to the
<code>data/</code> folder, or you can use the R code below to download
it directly to the folder.</p>
<div class="codewrapper sourceCode" id="cb153">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot1.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot1.csv"</span><span class="op">)</span></span>
<span><span class="va">annot1</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 Ã 2
   gene    gene_description                                                     
   &lt;chr&gt;   &lt;chr&gt;                                                                
 1 Cyp2d22 cytochrome P450, family 2, subfamily d, polypeptide 22 [Source:MGI Sâ¦
 2 Klk6    kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGI:1343166]   
 3 Fcrls   Fc receptor-like S, scavenger receptor [Source:MGI Symbol;Acc:MGI:19â¦
 4 Plp1    proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGI:97623]     
 5 Exd2    exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;Acc:MGI:192â¦
 6 Apod    apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]                   
 7 Gnb4    guanine nucleotide binding protein (G protein), beta 4 [Source:MGI Sâ¦
 8 Slc2a4  solute carrier family 2 (facilitated glucose transporter), member 4 â¦
 9 Asl     argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084]            
10 Gjc2    gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:2153060]    </code></pre>
</div>
<p>We now want to join these two tables into a single one containing all
variables using the <code>full_join()</code> function from the
<code>dplyr</code> package. The function will automatically find the
common variable to match columns from the first and second table. In
this case, <code>gene</code> is the common variable. Such variables are
called keys. Keys are used to match observations across different
tables.</p>
<div class="codewrapper sourceCode" id="cb155">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot1</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in full_join(rna_mini, annot1): object 'rna_mini' not found</code></pre>
</div>
<p>In real life, gene annotations are sometimes labelled
differently.</p>
<p>The <code>annot2</code> table is exactly the same than
<code>annot1</code> except that the variable containing gene names is
labelled differently. Again, either <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv" class="external-link">download
annot2.csv</a> yourself and move it to <code>data/</code> or use the R
code below.</p>
<div class="codewrapper sourceCode" id="cb157">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">download.file</span><span class="op">(</span>url <span class="op">=</span> <span class="st">"https://carpentries-incubator.github.io/bioc-intro/data/annot2.csv"</span>,</span>
<span>              destfile <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span>file <span class="op">=</span> <span class="st">"data/annot2.csv"</span><span class="op">)</span></span>
<span><span class="va">annot2</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">OUTPUT<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="output" tabindex="0"><code># A tibble: 10 Ã 2
   external_gene_name description                                               
   &lt;chr&gt;              &lt;chr&gt;                                                     
 1 Cyp2d22            cytochrome P450, family 2, subfamily d, polypeptide 22 [Sâ¦
 2 Klk6               kallikrein related-peptidase 6 [Source:MGI Symbol;Acc:MGIâ¦
 3 Fcrls              Fc receptor-like S, scavenger receptor [Source:MGI Symbolâ¦
 4 Plp1               proteolipid protein (myelin) 1 [Source:MGI Symbol;Acc:MGIâ¦
 5 Exd2               exonuclease 3'-5' domain containing 2 [Source:MGI Symbol;â¦
 6 Apod               apolipoprotein D [Source:MGI Symbol;Acc:MGI:88056]        
 7 Gnb4               guanine nucleotide binding protein (G protein), beta 4 [Sâ¦
 8 Slc2a4             solute carrier family 2 (facilitated glucose transporter)â¦
 9 Asl                argininosuccinate lyase [Source:MGI Symbol;Acc:MGI:88084] 
10 Gjc2               gap junction protein, gamma 2 [Source:MGI Symbol;Acc:MGI:â¦</code></pre>
</div>
<p>In case none of the variable names match, we can set manually the
variables to use for the matching. These variables can be set using the
<code>by</code> argument, as shown below with <code>rna_mini</code> and
<code>annot2</code> tables.</p>
<div class="codewrapper sourceCode" id="cb159">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot2</span>, by <span class="op">=</span> <span class="fu">c</span><span class="op">(</span><span class="st">"gene"</span> <span class="op">=</span> <span class="st">"external_gene_name"</span><span class="op">)</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in full_join(rna_mini, annot2, by = c(gene = "external_gene_name")): object 'rna_mini' not found</code></pre>
</div>
<p>As can be seen above, the variable name of the first table is
retained in the joined one.</p>
<div id="challenge9" class="callout challenge">
<div class="callout-square">
<i class="callout-icon" data-feather="zap"></i>
</div>
<div id="challenge-4" class="callout-inner">
<h3 class="callout-title">Challenge:<a class="anchor" aria-label="anchor" href="#challenge9"></a>
</h3>
<div class="callout-content">
<p>Download the <code>annot3</code> table by clicking <a href="https://carpentries-incubator.github.io/bioc-intro/data/annot3.csv" class="external-link">here</a>
and put the table in your data/ repository. Using the
<code>full_join()</code> function, join tables <code>rna_mini</code> and
<code>annot3</code>. What has happened for genes <em>Klk6</em>,
<em>mt-Tf</em>, <em>mt-Rnr1</em>, <em>mt-Tv</em>, <em>mt-Rnr2</em>, and
<em>mt-Tl1</em> ?</p>
</div>
</div>
</div>
<div id="accordionSolution9" class="accordion challenge-accordion accordion-flush">
<div class="accordion-item">
<button class="accordion-button solution-button collapsed" type="button" data-bs-toggle="collapse" data-bs-target="#collapseSolution9" aria-expanded="false" aria-controls="collapseSolution9">
  <h4 class="accordion-header" id="headingSolution9">
  Show me the solution
  </h4>
</button>
<div id="collapseSolution9" class="accordion-collapse collapse" data-bs-parent="#accordionSolution9" aria-labelledby="headingSolution9">
<div class="accordion-body">
<div class="codewrapper sourceCode" id="cb161">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">annot3</span> <span class="op">&lt;-</span> <span class="fu">read_csv</span><span class="op">(</span><span class="st">"data/annot3.csv"</span><span class="op">)</span></span>
<span><span class="fu">full_join</span><span class="op">(</span><span class="va">rna_mini</span>, <span class="va">annot3</span><span class="op">)</span></span></code></pre>
</div>
<div class="codewrapper">
<h3 class="code-label">ERROR<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="error" tabindex="0"><code>Error in full_join(rna_mini, annot3): object 'rna_mini' not found</code></pre>
</div>
<p>Genes <em>Klk6</em> is only present in <code>rna_mini</code>, while
genes <em>mt-Tf</em>, <em>mt-Rnr1</em>, <em>mt-Tv</em>,
<em>mt-Rnr2</em>, and <em>mt-Tl1</em> are only present in
<code>annot3</code> table. Their respective values for the variables of
the table have been encoded as missing.</p>
</div>
</div>
</div>
</div>
</section><section id="exporting-data"><h2 class="section-heading">Exporting data<a class="anchor" aria-label="anchor" href="#exporting-data"></a>
</h2>
<hr class="half-width"><p>Now that you have learned how to use <code>dplyr</code> to extract
information from or summarise your raw data, you may want to export
these new data sets to share them with your collaborators or for
archival.</p>
<p>Similar to the <code>read_csv()</code> function used for reading CSV
files into R, there is a <code>write_csv()</code> function that
generates CSV files from data frames.</p>
<p>Before using <code>write_csv()</code>, we are going to create a new
folder, <code>data_output</code>, in our working directory that will
store this generated dataset. We donât want to write generated datasets
in the same directory as our raw data. Itâs good practice to keep them
separate. The <code>data</code> folder should only contain the raw,
unaltered data, and should be left alone to make sure we donât delete or
modify it. In contrast, our script will generate the contents of the
<code>data_output</code> directory, so even if the files it contains are
deleted, we can always re-generate them.</p>
<p>Letâs use <code>write_csv()</code> to save the rna_wide table that we
have created previously.</p>
<div class="codewrapper sourceCode" id="cb163">
<h3 class="code-label">R<i aria-hidden="true" data-feather="chevron-left"></i><i aria-hidden="true" data-feather="chevron-right"></i>
</h3>
<pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">write_csv</span><span class="op">(</span><span class="va">rna_wide</span>, file <span class="op">=</span> <span class="st">"data_output/rna_wide.csv"</span><span class="op">)</span></span></code></pre>
</div>
<div id="keypoints1" class="callout keypoints">
<div class="callout-square">
<i class="callout-icon" data-feather="key"></i>
</div>
<div class="callout-inner">
<h3 class="callout-title">Keypoints<a class="anchor" aria-label="anchor" href="#keypoints1"></a>
</h3>
<div class="callout-content">
<ul><li>Tabular data in R using the tidyverse meta-package</li>
</ul></div>
</div>
</div>
</section></div> <!-- / div.lesson-content -->
    </main><!-- / main#main-content.main-content --><nav class="bottom-pagination mx-md-4" aria-label="Previous and Next Chapter"><div class="d-block d-sm-block d-md-none">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html">Next<i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
      <!-- content for large screens -->
      <div class="d-none d-sm-none d-md-block">
        <a class="chapter-link" href="../instructor/25-starting-with-data.html" rel="prev"><i aria-hidden="true" class="small-arrow" data-feather="arrow-left"></i>Previous: Starting with data</a>
        <a class="chapter-link float-end" href="../instructor/40-visualization.html" rel="next">Next: Data visualization... <i aria-hidden="true" class="small-arrow" data-feather="arrow-right"></i></a>
      </div>
    </nav></div> <!-- / div.primary-content.col-xs-12 -->
<!-- END:   inst/pkgdown/templates/content-instructor.html-->

      </div><!--/div.row-->
      		<footer class="row footer mx-md-3"><hr><div class="col-md-6">
				<p>This lesson is subject to the <a href="CODE_OF_CONDUCT.html">Code of Conduct</a></p>
        <p>
        
        <a href="https://github.com/carpentries-incubator/bioc-intro/edit/main/episodes/30-dplyr.Rmd" class="external-link">Edit on GitHub</a>
        
	
        | <a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CONTRIBUTING.md" class="external-link">Contributing</a> 
        | <a href="https://github.com/carpentries-incubator/bioc-intro/" class="external-link">Source</a></p>
				<p><a href="https://github.com/carpentries-incubator/bioc-intro/blob/main/CITATION" class="external-link">Cite</a> | <a href="mailto:laurent.gatto@uclouvain.be">Contact</a> | <a href="https://carpentries.org/about/" class="external-link">About</a></p>
			</div>
			<div class="col-md-6">
        
        <p>Materials licensed under <a href="../LICENSE.html">CC-BY 4.0</a> by the authors</p>
        
        <p><a href="https://creativecommons.org/licenses/by-sa/4.0/" class="external-link">Template licensed under CC-BY 4.0</a> by <a href="https://carpentries.org" class="external-link">The Carpentries</a></p>
        <p>Built with <a href="https://github.com/carpentries/sandpaper/tree/0.11.13" class="external-link">sandpaper (0.11.13)</a>,
        <a href="https://github.com/carpentries/pegboard/tree/0.5.1" class="external-link">pegboard (0.5.1)</a>,
      and <a href="https://github.com/carpentries/varnish/tree/0.2.15" class="external-link">varnish (0.2.15)</a>.</p>
			</div>
		</footer></div> <!-- / div.container -->
	<div id="to-top">
		<a href="#top">
			<i class="search-icon" data-feather="arrow-up" role="img" aria-label="Back to top"></i><br><span class="d-none d-sm-none d-md-none d-lg-none d-xl-block">Back</span> To Top
		</a>
	</div>
  <script type="application/ld+json">
    {
  "@context": "https://schema.org",
  "@type": "TrainingMaterial",
  "@id": "https://carpentries-incubator.github.io/bioc-introinstructor/30-dplyr.html",
  "dct:conformsTo": "https://bioschemas.org/profiles/TrainingMaterial/1.0-RELEASE",
  "description": "A Carpentries Lesson teaching foundational data and coding skills to researchers worldwide",
  "keywords": "software, data, lesson, The Carpentries",
  "name": "Manipulating and analysing data with dplyr",
  "creativeWorkStatus": "active",
  "url": "https://carpentries-incubator.github.io/bioc-introinstructor/30-dplyr.html",
  "identifier": "https://carpentries-incubator.github.io/bioc-introinstructor/30-dplyr.html",
  "dateCreated": "2020-09-14",
  "dateModified": "2023-04-03",
  "datePublished": "2023-04-03"
}

  </script><script>
		feather.replace();
	</script><!-- Matomo
    2022-11-07: we have gotten a notification that we have an overage for our
    tracking and I'm pretty sure this has to do with Workbench usage.
    Considering that I am not _currently_ using this tracking because I do not
    yet know how to access the data, I am turning this off for now.
  <script>
    var _paq = window._paq = window._paq || [];
    /* tracker methods like "setCustomDimension" should be called before "trackPageView" */
    _paq.push(["setDocumentTitle", document.domain + "/" + document.title]);
    _paq.push(["setDomains", ["*.preview.carpentries.org","*.datacarpentry.github.io","*.datacarpentry.org","*.librarycarpentry.github.io","*.librarycarpentry.org","*.swcarpentry.github.io", "*.carpentries.github.io"]]);
    _paq.push(["setDoNotTrack", true]);
    _paq.push(["disableCookies"]);
    _paq.push(['trackPageView']);
    _paq.push(['enableLinkTracking']);
    (function() {
          var u="https://carpentries.matomo.cloud/";
          _paq.push(['setTrackerUrl', u+'matomo.php']);
          _paq.push(['setSiteId', '1']);
          var d=document, g=d.createElement('script'), s=d.getElementsByTagName('script')[0];
          g.async=true; g.src='https://cdn.matomo.cloud/carpentries.matomo.cloud/matomo.js'; s.parentNode.insertBefore(g,s);
        })();
  </script>
  End Matomo Code --></body></html><!-- END:   inst/pkgdown/templates/layout.html-->

